provider/oci/common/client.go

```diff
@@ -0,0 +1,198 @@
+	_, err = o.ListCompartments(ctx, request)
+	return err
+}
+
+func (o *ociClient) GetInstanceVnicAttachments(instanceID instance.Id, compartmentID *string) (ociCore.ListVnicAttachmentsResponse, error) {
+	instID := string(instanceID)
+	request := ociCore.ListVnicAttachmentsRequest{
+		CompartmentId: compartmentID,
+		InstanceId:    &instID,
+	}
+	response, err := o.ListVnicAttachments(context.Background(), request)
+	if err != nil {
+		return ociCore.ListVnicAttachmentsResponse{}, errors.Trace(err)
+	}
+	return response, nil
+}
+
+func (o *ociClient) GetInstanceVnics(vnics []ociCore.VnicAttachment) ([]ociCore.GetVnicResponse, error) {
+	result := []ociCore.GetVnicResponse{}
+
+	for _, val := range vnics {
+		vnicID := val.VnicId
+		request := ociCore.GetVnicRequest{
+			VnicId: vnicID,
+		}
+		response, err := o.GetVnic(context.Background(), request)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		result = append(result, response)
+	}
+	return result, nil
+}
+
+func (o *ociClient) GetInstanceAddresses(instanceID instance.Id, compartmentID *string) ([]network.Address, error) {
+	attachments, err := o.GetInstanceVnicAttachments(instanceID, compartmentID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	vnics, err := o.GetInstanceVnics(attachments.Items)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	addresses := []network.Address{}
+
+	for _, val := range vnics {
+		if val.Vnic.PrivateIp != nil {
+			privateAddress := network.Address{
+				Value: *val.Vnic.PrivateIp,
+				Type:  network.IPv4Address,
+				Scope: network.ScopeCloudLocal,
+			}
+			addresses = append(addresses, privateAddress)
+		}
+		if val.Vnic.PublicIp != nil {
+			publicAddress := network.Address{
+				Value: *val.Vnic.PublicIp,
+				Type:  network.IPv4Address,
+				Scope: network.ScopePublic,
+			}
+			addresses = append(addresses, publicAddress)
+		}
+	}
+	return addresses, nil
+}
+
+func (j jujuConfigProvider) TenancyOCID() (string, error) {
+	if j.tenancyOCID == "" {
+		return "", errors.Errorf("tenancyOCID is not set")
+	}
+	return j.tenancyOCID, nil
+}
+
+func (j jujuConfigProvider) UserOCID() (string, error) {
+	if j.userOCID == "" {
+		return "", errors.Errorf("userOCID is not set")
+	}
+	return j.userOCID, nil
+}
+
+func (j jujuConfigProvider) KeyFingerprint() (string, error) {
+	if j.keyFingerprint == "" {
+		return "", errors.Errorf("keyFingerprint is not set")
+	}
+	return j.keyFingerprint, nil
+}
+
+func (j jujuConfigProvider) Region() (string, error) {
+	return j.region, nil
+}
+
+func (j jujuConfigProvider) PrivateRSAKey() (*rsa.PrivateKey, error) {
+	if j.key == nil {
+		return nil, errors.Errorf("private key is not set")
+	}
+
+	key, err := ociCommon.PrivateKeyFromBytes(
+		j.key, &j.passphrase)
+	return key, err
+}
+
+func (j jujuConfigProvider) KeyID() (string, error) {
+	if j.tenancyOCID == "" || j.userOCID == "" || j.keyFingerprint == "" {
+		return "", errors.Errorf("config provider is not properly initialized")
```

Comment:
This sort of check can be moved into the Validate() code instead of waiting until we need the KeyID to know if we have valid configuration.

Reply:
Good point.

Reply:
The function does however have an error object return, so we can do it in a validate (whici is not part of the ```ociCommon.ConfigurationProvider``` interface), and render that field useless.